org.eclipse.tm4e.core/src/main/java/org/eclipse/tm4e/core/model/AbstractTMModel.java

```diff
@@ -128,170 +70,203 @@ public int getSize() {
 		 * @param name
 		 *            the thread's name
 		 */
-		public BackgroundThread(String name) {
+		public TokenizerThread(String name, AbstractTMModel model) {
 			super(name);
+			this.model = model;
+			this.tokenizer = new Tokenizer(model.getGrammar());
 			setPriority(Thread.MIN_PRIORITY);
 			setDaemon(true);
 		}
 
-		/**
-		 * Suspends the caller of this method until this background thread has
-		 * emptied the dirty region queue.
-		 */
-		public void suspendCallerWhileDirty() {
-			boolean isDirty;
-			do {
-				synchronized (fDirtyRegionQueue) {
-					isDirty = fDirtyRegionQueue.getSize() > 0;
-					if (isDirty) {
+		@Override
+		public void run() {
+			if (interrupted())
+				return;
+
+			// initialize
+			int linesLength = model.getNumberOfLines();
+			for (int line = 0; line < linesLength; line++) {
+				try {
+					model.lines.addLine(line);
+				} catch (Exception e) {
+					e.printStackTrace();
+				}
+			}
+			model.lines.forEach(ModelLine::resetTokenizationState);
+			model.lines.get(0).setState(tokenizer.getInitialState());
+			model._invalidLineStartIndex = 0;
+			
+			while (!interrupted()) {
+				if (model.lines.getSize() == 0) {
+					synchronized (model.lines) {
 						try {
-							fDirtyRegionQueue.wait();
+							model.lines.wait(500); // release synchronized lock at least every 500ms
 						} catch (InterruptedException x) {
 						}
 					}
 				}
-			} while (isDirty);
-		}
-
-		/**
-		 * Reset the background thread as the text viewer has been changed,
-		 */
-		public void reset() {
+				while (model.lines.getSize() >= 0 && model._invalidLineStartIndex < model.lines.getSize()) {

```

Comment:
When I do open a big file and I do nothing my CPU is growing although I do nothing.

After studying the problem, it seems that when a Java list is big the List#size() can consume a lot CPU.

I have done a simple test:

```
		List<String> list = new ArrayList<>();
		for (int i = 0; i < 999; i++) {
			list.add("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee");
		}
		
		Thread t = new Thread() {
			@Override
			public void run() {
				while(true) {
					// Uncomment the wait code to avoid CPU problem
					/* synchronized (list) {
						try {
							list.wait(1);
						} catch (InterruptedException e) {
						}
					}*/
					int size = list.size();					
				}
			}
		};
		t.start();
		t.join();
```

With my Windows OS, my CPU is working, but if I uncomment synchronized block which use wait, I have not this problem.

In other words I think we should use wait to avoid computing list size every time and causes CPU problems